---
title: "R Notebook"
author: Findlay Bewicke-Copley PhD
output: html_notebook
---

```{r setup, echo=FALSE}
# Packages to load
packages<- c("tidyverse", "patchwork", "Biobase","limma","pamr","survival",
             "survminer", "reshape2", "xlsx", "ggplotify", "gridExtra","ggalluvial",
             "PCAtools", "ggvenn", "gtools")
## Load or install these packages
for ( package in packages ) {
    if (! suppressPackageStartupMessages(require(package, character.only = TRUE)) ) {
        BiocManager::install(package,update = FALSE)
        require(package, character.only = TRUE)
    }
}
## Function to calculate the Linear Predictor 
GenerateLinearPredictor <- function( betas , ExprSet, scale = TRUE, cutoff = FALSE) {
        ## filter the betas down to only those in ExprSet
        betas <- betas[names(betas) %in% featureNames(ExprSet)]
        ## Filter the ExprSet down to the remaining beta genes
        FinObject <- ExprSet[names(betas),]
        ## Store betas
        fData(FinObject)$betas <- betas
        ## It's important that these are now ordered exactly the same.
        ## Multiply the expression of the genes by the beta values.
        FinObject$LinearPredictor <- colSums(sweep(exprs(FinObject),
                                                   MARGIN=1,
                                                   fData(FinObject)$betas,`*`))
        if ( scale ) {
            ## scale generates a zscore
            FinObject$LinearPredictor <- scale(FinObject$LinearPredictor)
        }
        ## Run a quick look taking the median value
        FinObject$MedianGroups <- ifelse(FinObject$LinearPredictor > median(FinObject$LinearPredictor), "High Risk", "Low Risk")
        ## Split groups based on the cut off supplied
        if ( is.numeric(cutoff) ) {
            FinObject$BestGroup <- cut(FinObject$LinearPredictor, breaks = c(-Inf, cutoff , Inf), labels = c("Low", "High"))
        }
        FinObject
}
theme_set(theme_classic(base_size = 10,
                            base_family = 'helvetica'))
Colours <- c("Diagnosis" = "#70ad47",
             "Relapse" = "#a50021",
             "ABC"="#5b9bd5",
             "GCB"="#ed7d31",
             "UNC"="#b5b3b3",
             "Unclassified"="#b5b3b3"
             )
Cohort_Colours <- c("Diagnosis" = "#70ad47",
                    "Relapse" = "#a50021",
                    "ABC"="#5b9bd5",
                    "GCB"="#ed7d31",
                    "Unclassified"="#b5b3b3",
                    "Nodal" = "darkorchid4",
                    "Extra-nodal" = "darkgoldenrod2",
                    "NA" = "Black",
                    ">2y" = "#806000",
                    "<2y" = "#ffd966",
                    "M" = "#e4d882",
                    "F" = "#e29b7b",
                    "<50" = "#808000",
                    "50-59" = "#666633",
                    "60-69" = "#996633",
                    ">70" = "#663300",
                    "PR" = "#7030a0",
                    "CR" = "#a9d08e")
```

## Pairs data

These data were put together in the DLBCL_Paired_Relapse_Data_Creation.Rmd notebook

```{r}
## Load Data
Data <- list()
## This has been slightly updated
Data$PairsData <- readRDS("RDS/rrDLBCL.batchCorrected.eSet.rds")
pData(Data$PairsData)
summary(Data$PairsData$Relapse_Time)

```

# Table 1

## Cohort Info

I want a few informative stats from the data for a table

```{r eval=FALSE}
CohortData <- list(
    Age = subset(pData(Data$PairsData), Timepoint == "Diagnosis", select = Age) %>%
        summary(),
    TimeToRelapse = subset(pData(Data$PairsData), Timepoint == "Diagnosis", select = Relapse_Time) %>%
        summary(), 
    Sex = subset(pData(Data$PairsData), Timepoint == "Diagnosis", select = Sex) %>%
        table(),
    COO_Diagnosis = subset(pData(Data$PairsData), Timepoint == "Diagnosis", select = COO) %>%
        table(),
    COO_Relase = subset(pData(Data$PairsData), Timepoint == "Relapse", select = COO) %>%
        table(),
    COO_Stable = {x <- pivot_wider(
        pData(Data$PairsData), id_cols = PatID, names_from = Timepoint, values_from = c("COO"))
    cbind(ABC = sum(x$Diagnosis == "ABC" & x$Relapse == "ABC"),
          GCB = sum(x$Diagnosis == "GCB" & x$Relapse == "GCB"),
          UNC = sum(x$Diagnosis == "UNC" & x$Relapse == "UNC")) },
    Site_Diagnosis = subset(pData(Data$PairsData), Timepoint == "Diagnosis", select = Nodal) %>%
        table(),
    Site_Relase = subset(pData(Data$PairsData), Timepoint == "Relapse", select = Nodal) %>%
        table(),
    Site_Stable = {x <- pivot_wider(
        pData(Data$PairsData), id_cols = PatID, names_from = Timepoint, values_from = c("Nodal"))
    cbind(Nodal = sum(x$Diagnosis == TRUE & x$Relapse == TRUE),
          ExtraNodal = sum(x$Diagnosis != TRUE & x$Relapse != TRUE))},
    TreatmentResponse = subset(pData(Data$PairsData), Timepoint == "Diagnosis", select = Response) %>%
        table()
)
```



# Figure 1

## Cohort Information

Creating a plot of the characteristics of the Cohort 

```{r}
Cohort <- rbind(cbind(pData(Data$PairsData)["Timepoint"],"Timepoint") %>% 
                    setNames(c("val","yaxis")) %>% 
                    rownames_to_column(var="sample"),
                cbind(pData(Data$PairsData)["COO"],"COO") %>% 
                    setNames(c("val", "yaxis")) %>% 
                    rownames_to_column(var="sample"),
                cbind(ifelse(pData(Data$PairsData)["Relapse_Time"] > 2,"<2y",">2y"),"Relapse Time") %>% as.data.frame() %>%
                    setNames(c("val", "yaxis")) %>%
                    rownames_to_column(var="sample"),
                cbind(pData(Data$PairsData)["Response"],"Response") %>% 
                    setNames(c("val", "yaxis")) %>% 
                    rownames_to_column(var="sample"),
                cbind(pData(Data$PairsData) %>% 
                          subset(.,Timepoint == "Diagnosis") %>% 
                          .["Age"] %>% unlist() %>%
                          lapply(.,function(x) rep(x,2)) %>% 
                          unlist() %>%
                          as.numeric() %>% 
                          cut(.,breaks=c(0,50,60,70,Inf),
                              labels=c("<50","50-59","60-69",">70")) %>%
                          as.data.frame() %>%
                          'rownames<-'(sampleNames(Data$PairsData)),"Age") %>% 
                    setNames(c("val", "yaxis")) %>%
                    rownames_to_column(var="sample"),
                cbind(ifelse(pData(Data$PairsData)["Nodal"] == TRUE, "Nodal", "Extra-nodal") %>% as.data.frame(),
                      "Nodal/Extra Nodal") %>% 
                    setNames(c("val", "yaxis")) %>%
                    rownames_to_column(var="sample"),
                cbind(pData(Data$PairsData)["Sex"],"Gender") %>%
                    setNames(c("val", "yaxis")) %>% 
                    rownames_to_column(var="sample")) %>%
    ggplot(data=.,aes(x=sample, y=factor(yaxis,
                                         levels=rev(c("Timepoint","COO","Nodal/Extra Nodal","Relapse Time","Gender","Age","Response"))),
                      fill=factor(val, levels=names(Cohort_Colours)))) +
    labs(y="", x="") +
    geom_tile() + 
    scale_fill_manual(values=Cohort_Colours, name = "" ) +
    geom_hline(yintercept=seq(0.5,1.5+8,1), color="white") +
    geom_segment(data=as.data.frame(cbind(x = seq(1.5,76.5,1),
                                          xend = seq(1.5,76.5,1),
                                          y = rep(c(4.52, 0.50), 38),
                                          yend = rep(7.48, 76))), 
                 aes(x=x, y=y, xend=xend,yend=yend), inherit.aes = FALSE, colour="white") +
    theme(axis.text.x = element_blank(),
          axis.ticks = element_blank(),
          axis.line = element_blank(),
          legend.position = "none")
CohortLegTheme <- theme(legend.text=element_text(size=7),
                        legend.title=element_text(size=7, face="bold"),
                        legend.key.size = unit(5, "mm"),
                        legend.position = "bottom",
                        legend.direction = "vertical",
                        legend.box.just = "top",
                        legend.background = element_blank()) 
timepointKey <- get_legend(
    ggplot() +
        geom_raster(
            data = Cohort_Colours[1:2] %>% as.data.frame() %>% rownames_to_column("Name"),
            aes(
                x = 1,
                y = factor(Name, levels = rev(Name)),
                fill = Name
            )
        ) +
        scale_fill_manual(values = Cohort_Colours[1:2], name = "Timepoint") + CohortLegTheme
)
cooKey <- get_legend(
    ggplot() +
        geom_raster(
            data = Cohort_Colours[3:5] %>% as.data.frame() %>% rownames_to_column("Name"),
            aes(
                x = 1,
                y = factor(Name, levels = rev(Name)),
                fill = Name
            )
        ) +
        scale_fill_manual(values = Cohort_Colours[3:5], name = "COO") + CohortLegTheme
)
nodeKey <- get_legend(
    ggplot() +
        geom_raster(
            data = Cohort_Colours[6:7] %>% as.data.frame() %>% rownames_to_column("Name"),
            aes(
                x = 1,
                y = factor(Name, levels = rev(Name)),
                fill = Name
            )
        ) +
        scale_fill_manual(values = Cohort_Colours[6:7], name = "Nodal/Extra Nodal")  + CohortLegTheme
)
t2rKey <- get_legend(
    ggplot() +
        geom_raster(
            data = Cohort_Colours[9:10] %>% as.data.frame() %>% rownames_to_column("Name"),
            aes(
                x = 1,
                y = factor(Name, levels = rev(Name)),
                fill = Name
            )
        ) +
        scale_fill_manual(values = Cohort_Colours[9:10], name = "Relapse Time") + CohortLegTheme
)


genderKey <- get_legend(
    ggplot() +
        geom_raster(
            data = Cohort_Colours[11:12] %>% as.data.frame() %>% rownames_to_column("Name"),
            aes(
                x = 1,
                y = factor(Name, levels = rev(Name)),
                fill = Name
            )
        ) +
        scale_fill_manual(values = Cohort_Colours[11:12], name = "Gender") + CohortLegTheme
)
ageKey <- get_legend(
    ggplot() +
        geom_raster(
            data = Cohort_Colours[13:16] %>% as.data.frame() %>% rownames_to_column("Name"),
            aes(
                x = 1,
                y = factor(Name, levels = rev(Name)),
                fill = Name
            )
        ) +
        scale_fill_manual(values = Cohort_Colours[13:16], name = "Age") + CohortLegTheme
)
responseKey <- get_legend(
    ggplot() +
        geom_raster(
            data = Cohort_Colours[17:18] %>% as.data.frame() %>% rownames_to_column("Name"),
            aes(
                x = 1,
                y = factor(Name, levels = rev(Name)),
                fill = Name
            )
        ) +
        scale_fill_manual(values = Cohort_Colours[17:18], name = "Response") + CohortLegTheme
)
Cohort <-
    Cohort + (as.ggplot(
        gtable_cbind(
            timepointKey,
            cooKey,
            nodeKey,
            t2rKey,
            genderKey,
            ageKey,
            responseKey
        )
    )) + plot_layout(heights = c(4, 1))
```

## Alluvial Diagram of COO

```{r message=FALSE, echo=FALSE, warning=FALSE}
## COO stability
COOStability <- spread(pData(Data$PairsData)[,c("Timepoint","COO","PatID")], Timepoint, COO )
COOStability$Diagnosis <- ifelse(COOStability$Diagnosis == "Unclassified", "UNC", COOStability$Diagnosis %>% as.character())
COOStability$Relapse <- ifelse(COOStability$Relapse == "Unclassified", "UNC", COOStability$Relapse %>% as.character())
## Diag-Relapse COO data.frame
COOStability <- melt(table(COOStability[,c("Diagnosis","Relapse")]))
## Set the levels so that the UNC sits between ABC and GCB
COOStability$Diagnosis <- factor(COOStability$Diagnosis, levels=c("ABC","UNC","GCB"))
COO <- ggplot(subset(COOStability, value != 0),
              aes(y=value, axis1 = Diagnosis, axis2=Relapse)) +
    ## Fill with the COO at diagnosis.
    geom_alluvium(aes(fill = Diagnosis, width=1/12)) +
    geom_stratum(width = 1/12, alpha = .05) +
    scale_fill_manual(values=Colours) +
    scale_x_discrete(limits = c("Diagnosis", "Relapse"), expand = c(.05, .05)) +
    theme(axis.line.y = element_blank(),
          axis.ticks.y = element_blank(),
          axis.title.y = element_blank(),
          axis.text.y = element_blank(),
          axis.text.x = element_text(size=12),
          legend.position = "none") +
        geom_label(stat = "stratum", aes(label = after_stat(stratum)))
```

## PCA plot of normalised batch corrected Data

```{r}
pca <- pca(exprs(Data$PairsData), metadata = pData(Data$PairsData))
PCA_Timepoint <- biplot(pca,
                        colby = "Timepoint",
                        gridlines.major = FALSE, 
                        gridlines.minor = FALSE,
                        pointSize = 3, 
                        lab= FALSE ) +
        scale_color_manual(name="",values=Colours[1:2]) +
        theme_classic() +
        theme(legend.position = "bottom",
              legend.direction = "horizontal")


PCA_COO <- biplot(pca,
                  colby = "COO", 
                  gridlines.major = FALSE, gridlines.minor = FALSE,
                  pointSize = 3, 
                  lab= FALSE ) +
        scale_color_manual(name="",values=Colours[3:5]) +
        theme_classic() +
        theme(legend.position = "bottom",
              legend.direction = "horizontal")
PCA <- (PCA_Timepoint + PCA_COO )
```

## GSEA Venn Diagrams from DEG

```{r}
ABC <- Data$PairsData[,Data$PairsData$COO == "ABC"]
## Build the design, with diagnosis as the intercept.
## Pair samples by PatID in the design
## Then assess the differences at relapse (Timepoint)
ABC_design <- model.matrix(~as.factor(as.vector(ABC$PatID))+ABC$Timepoint)
## DGE
ABC_fit <- lmFit(exprs(ABC), ABC_design)
ABC_fit <- eBayes(ABC_fit)
ABC_FoldChange <- topTable(ABC_fit,
                           n=Inf,
                           coef=ncol(ABC_design))
## GCB analysis
GCB <- Data$PairsData[,Data$PairsData$COO == "GCB"]
## Build the design, with diagnosis as the intercept.
## Pair samples by PatID in the design
## Then assess the differences at relapse (Timepoint)
GCB_design <- model.matrix(~as.factor(as.vector(GCB$PatID))+GCB$Timepoint)
## DGE
GCB_fit <- lmFit(exprs(GCB), GCB_design)
GCB_fit <- eBayes(GCB_fit)
GCB_FoldChange <- topTable(GCB_fit,
                           n=Inf,
                           coef=ncol(GCB_design))
## Export genes for GSEA, using the t_statistic as the ranking variable
cbind(rownames(ABC_FoldChange),
      ABC_FoldChange$t) %>% 
        as.data.frame() %>%
        setNames(c("#Probe Set ID","t_statistic")) %>%
        arrange(., t_statistic %>% as.character() %>% as.numeric() %>% -.) %>%
        write.table("GSEA/ABC_Limma.rnk",
                    quote=FALSE,
                    row.names = FALSE,
                    sep="\t")
cbind(rownames(GCB_FoldChange),
      GCB_FoldChange$t) %>% 
        as.data.frame() %>%
        setNames(c("#Probe Set ID","t_statistic")) %>%
        arrange(., t_statistic %>% as.character() %>% as.numeric() %>% -.) %>%
        write.table("GSEA/GCB_Limma.rnk",
                    quote=FALSE,
                    row.names = FALSE,
                    sep="\t")
## Run GSEA using genepattern
## Kegg/Reactome
## Read in the results from gene pattern
ABC_GSEA <- rbind(read.delim("GSEA/ABC_t/gsea_report_for_na_pos_1581443460638.xls"),
                  read.delim("GSEA/ABC_t/gsea_report_for_na_neg_1581443460638.xls"))
GCB_GSEA <- rbind(read.delim("GSEA/GCB_t/gsea_report_for_na_pos_1581443631166.xls"),
                  read.delim("GSEA/GCB_t/gsea_report_for_na_neg_1581443631166.xls"))
## Print out the significant (FDR > 0.05) gene sets from the GSEA analysis, ordered by NES
ABC_GSEA %>% subset(FDR.q.val < 0.1) %>% arrange(- NES) %>% droplevels.data.frame() %>%  
        write.xlsx("GSEA/SigPathways.xlsx", sheetName="ABC")
## subset(GCB_GSEA, FDR.q.val <= 0.1 & NES > 0)
GCB_GSEA %>% subset(FDR.q.val < 0.1) %>% arrange(- NES) %>% droplevels.data.frame() %>%
        write.xlsx("GSEA/SigPathways.xlsx", sheetName="GCB", append=TRUE)
## Venn diagram of enriched gene sets
GSEA_Up_Venn <- ggvenn::ggvenn(list("ABC" = c(ABC_GSEA %>% subset(FDR.q.val <= 0.1 & NES > 0) %>% droplevels.data.frame() %>% .[,"NAME"]),
                                    "GCB" = c(GCB_GSEA %>% subset(FDR.q.val <= 0.1 & NES > 0) %>% droplevels.data.frame() %>% .[,"NAME"])),
                               fill_color = c("#5b9bd5","#ed7d31"), 
                               fill_alpha = 0.3,
                               stroke_color = "grey",
                               show_percentage = FALSE) +
        labs(title="Enriched at relapse")
GSEA_Down_Venn <- ggvenn::ggvenn(list("ABC" = c(ABC_GSEA %>% subset(FDR.q.val <= 0.1 & NES < 0) %>% droplevels.data.frame() %>% .[,"NAME"]),
                                      "GCB" = c(GCB_GSEA %>% subset(FDR.q.val <= 0.1 & NES < 0) %>% droplevels.data.frame() %>% .[,"NAME"])),
                                 fill_color = c("#5b9bd5","#ed7d31"), 
                                 fill_alpha = 0.3, 
                                 stroke_color = "grey",
                                 show_percentage = FALSE,) + 
        labs(title="Reduced at relapse")
```



## GSEA heatmap

```{r}
GSEA <- read.xlsx("GSEA/rrDLBCL_pairs_GSEA.xlsx", sheetName = "For Alex")
## chromosome maintenance, DNA repair, and rRNA processing 
ABC_GSEA <- subset(GSEA, NAME %in% c(
        "REACTOME_DNA_METHYLATION",
        "REACTOME_TELOMERE_MAINTENANCE",
        "REACTOME_ACTIVATED_PKN1_STIMULATES_TRANSCRIPTION_OF_AR_ANDROGEN_RECEPTOR_REGULATED_GENES_KLK2_AND_KLK3",
        "REACTOME_EUKARYOTIC_TRANSLATION_INITIATION",
        ## rRNA processing
        "KEGG_RIBOSOME",
        "REACTOME_RRNA_PROCESSING_IN_THE_NUCLEUS_AND_CYTOSOL",
        "REACTOME_RRNA_PROCESSING",
        ## DNA repair
        "KEGG_BASE_EXCISION_REPAIR",
        "REACTOME_NONSENSE_MEDIATED_DECAY_NMD_INDEPENDENT_OF_THE_EXON_JUNCTION_COMPLEX_EJC"
)) %>% 
        arrange(-NES.ABC) %>% droplevels()
ABC_GSEA_Heatmap <- ABC_GSEA %>%
        melt(id.vars="NAME", measure.vars=c("NES.ABC","NES.GCB")) %>%
        ggplot(aes(x = variable %>% gsub(".*\\.", "", .), 
                   y = NAME %>% gsub("KEGG_", "", .) %>%
                           gsub("REACTOME_", "", .) %>%
                           gsub("_", " ",.) %>%
                           str_to_sentence() %>% 
                           gsub("Dna", "DNA", .) %>%
                           gsub("Rrna", "rRNA", .) %>%
                           gsub("pkn1", "PKN1", .) %>%
                           gsub(" ar ", " AR ", .) %>%
                           gsub("klk", "KLK", .) %>%
                           gsub("ejc", "EJC", .) %>%
                           gsub("b cell", "B cell", .) %>%
                           gsub("bcr", "BCR", .) %>%
                           gsub("tcr", "TCR", .) %>%
                           str_wrap(width=30) %>% 
                           factor(., levels=rev(.[!duplicated(.)])),
                   fill = value)) + 
        geom_tile() +
        scale_fill_gradient2(name = "NES",
                             low="blue", 
                             mid="white",
                             high="red",
                             limits=c(-3,3), 
                             breaks=seq(-3,3,1.5)) +
        labs(x = "COO", y = "") + 
        geom_text(data = ABC_GSEA %>% 
                          melt(id.vars="NAME", 
                               measure.vars=c("FDR.q.val.ABC",
                                              "FDR.q.val.GCB")),
                  aes(label = value %>% stars.pval(.))
        )
GCB_GSEA <- subset(GSEA, NAME %in% c(
        ## Cytokine
        "KEGG_CHEMOKINE_SIGNALING_PATHWAY","REACTOME_INTERLEUKIN_1_FAMILY_SIGNALING","REACTOME_INTERLEUKIN_1_SIGNALING",
        "REACTOME_INTERLEUKIN_4_AND_INTERLEUKIN_13_SIGNALING",
        ## Antigen Processing
        "REACTOME_ANTIGEN_PROCESSING_CROSS_PRESENTATION","REACTOME_BINDING_AND_UPTAKE_OF_LIGANDS_BY_SCAVENGER_RECEPTORS",
        "REACTOME_CROSS_PRESENTATION_OF_SOLUBLE_EXOGENOUS_ANTIGENS_ENDOSOMES",
        ## Adaptive Immune System
        "REACTOME_DOWNSTREAM_SIGNALING_EVENTS_OF_B_CELL_RECEPTOR_BCR","REACTOME_DOWNSTREAM_TCR_SIGNALING",
        "REACTOME_SIGNALING_BY_THE_B_CELL_RECEPTOR_BCR")) %>% 
        arrange(-NES.GCB)
GCB_GSEA_Heatmap <- GCB_GSEA %>%
        melt(id.vars="NAME", measure.vars=c("NES.ABC","NES.GCB")) %>%
        ggplot(aes(x = variable %>% gsub(".*\\.", "", .),
                   y = NAME %>% gsub("KEGG_", "", .) %>%
                           gsub("REACTOME_", "", .) %>% 
                           gsub("_", " ",.) %>%
                           str_to_sentence() %>% 
                           gsub("Dna", "DNA", .) %>%
                           gsub("Rrna", "rRNA", .) %>%
                           gsub("pkn1", "PKN1", .) %>%
                           gsub(" ar ", " AR ", .) %>%
                           gsub("klk", "KLK", .) %>%
                           gsub("ejc", "EJC", .) %>%
                           gsub("b cell", "B cell", .) %>%
                           gsub("bcr", "BCR", .) %>%
                           gsub("tcr", "TCR", .) %>%
                           str_wrap(width=30) %>% 
                           factor(., levels=rev(.[!duplicated(.)])),
                   fill = value)) + 
        geom_tile() +
        scale_fill_gradient2(name = "NES",
                             low="blue", 
                             mid="white",
                             high="red",
                             limits=c(-3,3), 
                             breaks=seq(-3,3,1.5)) +
        labs(x = "COO", y = "") + 
        geom_text(data = GCB_GSEA %>% 
                          melt(id.vars="NAME", 
                               measure.vars=c("FDR.q.val.ABC",
                                              "FDR.q.val.GCB")),
                  aes(label = value %>% stars.pval(.))
        )

GESA_heatmap <- ( ABC_GSEA_Heatmap |  GCB_GSEA_Heatmap ) + plot_layout(guides="collect")
```


## PAMR for ABC genes

```{r}
library(pamr)
REDDY <- list()
REDDY$Data <- readRDS("RDS/ExternalDataset/Reddy.Processed.ABC.rds")
## Relaxed cut offs for PAMR genes
PAMR_Genes <- subset(ABC_FoldChange, P.Value <= 0.05)
## Return only the PAMR genes
REDDY$ABC <- REDDY$Data[fData(REDDY$Data)$external_gene_name %in%
                                rownames(PAMR_Genes),]
## Making a list for pamr. 
data=list(x=exprs(REDDY$ABC), 
          ## Survival Time (Years)
          survival.time=REDDY$ABC$OS_TIME,
          ## Censoring Status (Dead = 1 Alive = 0 (is this right?))
          censoring.status=REDDY$ABC$OS_IND,
          geneid=fData(REDDY$ABC)$external_gene_name)
## train pamr
a3 <- pamr.train(data, ngroup.survival=2)
## Cross-validation using pam.cv
mycv <- pamr.cv(a3,  data)
THRESHOLD=1.3
#make class predictions
REDDY$ABC$riskgroups <- pamr.predict(a3,data$x, threshold=THRESHOLD)
## Rename these groups 1 = High Risk, 2 = Low Risk
REDDY$ABC$group <- ifelse(REDDY$ABC$riskgroups == 1,"High Risk","Low Risk")

AnnotReddy <- data.frame(x=c(11.5,11.5),
                         y=c(0.25,0.45),
                         label=paste("n =", table(REDDY$ABC$group)))
## Using the PAMR defined groups draw a KM graph.
ReddyTrainSurv <- ggsurvplot(survfit(Surv(OS_TIME, OS_IND) ~ group,
                                     data=pData(REDDY$ABC)),
                             pval = TRUE,
                             pval.method = TRUE,
                             pval.size=4,
                             surv.median.line	= "hv",
                             ylab="Overall Survival")$plot +
        geom_text(data=AnnotReddy, aes(x=x, y=y, label=label), size=4) +
        theme(legend.position = "none",
              axis.text.x = element_text(size=10),
              axis.text.y = element_text(size=10),
              axis.title.x = element_text(size=12),
              axis.title.y = element_text(size=12))
```

### Gene forest plot

```{r}
#A function to list the genes that survive the thresholding, from the nearest shrunken centroid classi- fier produced by pamr.train
genes_sel <- pamr.listgenes(a3, data, threshold=THRESHOLD)
REDDY$ABC <- REDDY$ABC[fData(REDDY$ABC)$external_gene_name %in% genes_sel[,"id"],]
betas <- summary(coxph(Surv(REDDY$ABC$OS_TIME, REDDY$ABC$OS_IND) ~ ., 
                       data=data.frame(t(exprs(REDDY$ABC)))))$coefficients[,"coef"] %>% 
        setNames(fData(REDDY$ABC)$external_gene_name)

## Run a cox propotional hazards test to model the effect of the individual genes.
Gene_Cox <- summary(coxph(
        Surv(REDDY$ABC$OS_TIME,
             REDDY$ABC$OS_IND) ~ ., 
        ## the variables need to be the columns, so the data.frane needs rotated otherwise we'd get HR for the samples
        data=REDDY$ABC %>% exprs() %>% t() %>% as.data.frame() ))
## Extract all the coxph data for the genes.
GeneInfo <- cbind(
        ## This gives HR (expr) and confidence intervals for the genes.
        Gene_Cox$conf.int,
        ## This is the Wald test statistic
        Gene_Cox$coefficients[,5]) %>%
        as.data.frame(.) %>%
        setNames(c("exp","exp-coef","low95","up95","P"))
## Get gene names instead of ensID
rownames(GeneInfo) <- fData(REDDY$ABC[rownames(GeneInfo),])$external_gene_name
## Order the genes by the P value (Wald Statistic)
GeneInfo <- GeneInfo[order(-GeneInfo$P),]
GenePlot <- ggplot(GeneInfo,
                   ## need to re-factor the rownames for y and set the levels otherwise it'll reorder again ... 
                   aes(y=factor(rownames(GeneInfo), 
                                levels=rownames(GeneInfo[order(GeneInfo$exp),])),
                       x=exp,
                       xmin=GeneInfo$low95,
                       xmax=GeneInfo$up95)) +
    geom_text(aes(x=0),
              label=ifelse(GeneInfo$P > 0.1, "",
                           ifelse(GeneInfo$P > 0.05, ".",
                                  ifelse(GeneInfo$P > 0.01, "*","**"))), 
              hjust=0) +
    ## Label 1
    geom_vline(xintercept = 1, 
               col="grey") +
    ## Add error bars based on confidence intervals
    # geom_errorbarh(xmin=GeneInfo$low95,
    #                xmax=GeneInfo$up95) + 
    ## Add HR
    geom_pointrange(size=.5) +
    ## Change titles
    labs(x="HR (95%)", y="Genes")+
    ## Make it boring
    theme_classic() +
    ## Show everything
    expand_limits(x=c(0,4)) +
    ## Remove Y axis line
    theme(axis.line.y = element_blank(),
          axis.ticks.y = element_blank(),
          axis.text.y = element_text(size=7),
          axis.text = element_text(size=10),
          axis.title = element_text(size=12))
PAMR <- ReddyTrainSurv + GenePlot + plot_layout(widths=c(4,1))
saveRDS(betas, "RDS/ABC_predictor_betas.rds")
```


## Figure 1 panels

```{r}
if (! dir.exists("Figures/Paper_Figures") ) {dir.create("Figures/Paper_Figures")}
## A
Cohort
ggsave("Figures/Paper_Figures/Figure1-A-Cohort.svg", width=5, height=3)
## Legend made with w=9, h=5
## B
COO
ggsave("Figures/Paper_Figures/Figure1-B-COO.svg", width=5, height=3)
## C
PCA
ggsave("Figures/Paper_Figures/Figure1-C-PCA.svg", width=7, height=4)
## D
GSEA_Up_Venn
ggsave("Figures/Paper_Figures/Figure1-D-GSEA-UP.svg", width=5, height=3)
GSEA_Down_Venn
ggsave("Figures/Paper_Figures/Figure1-D-GSEA-DOWN.svg", width=5, height=3)
## E
GESA_heatmap <- ( ABC_GSEA_Heatmap |  GCB_GSEA_Heatmap ) + plot_layout(guides="collect")
ggsave("Figures/Paper_Figures/Figure1-E-GSEA_heatmap.svg", height=4, width = 8)
## F
PAMR
ggsave("Figures/Paper_Figures/Figure1-F-PAMR.svg", width=5, height=3)
```

## Supplemental Figure 1

```{r}
pca <- pca(exprs(Data$PairsData), metadata = pData(Data$PairsData) %>% 
             rownames_to_column %>%
             mutate(rel_time = ifelse(Relapse_Time > 2, ">2", "<=2")) %>%
             column_to_rownames())
Sup1A <- biplot(pca,colby = "PatID", 
                gridlines.major = FALSE, gridlines.minor = FALSE,
                pointSize = 3, 
                lab = NULL) +
        theme_classic() +
        theme(legend.position = "none")
Sup1B <- biplot(pca, colby = "rel_time", 
                gridlines.major = FALSE, gridlines.minor = FALSE,
                pointSize = 3, 
                lab= NULL) +
        theme_classic() +
        theme(legend.position = "bottom",
              legend.direction = "horizontal")
Sup1C <- biplot(pca, colby = "Nodal", 
                gridlines.major = FALSE, gridlines.minor = FALSE,
                pointSize = 3, 
                lab= NULL) +
        theme_classic() +
        theme(legend.position = "bottom",
              legend.direction = "horizontal")

Sup1D1 <- Volcano::volcanoPlot(ABC_FoldChange, SigName = "P.Value", FCName = "logFC",fcThres = FALSE, pThres = 0.05, PRINT = FALSE)
Sup1D2 <- Volcano::volcanoPlot(GCB_FoldChange, SigName = "P.Value", FCName = "logFC",fcThres = FALSE, pThres = 0.05, PRINT = FALSE)
## E
Sup1E <- ggvenn::ggvenn(list(ABC = subset(ABC_FoldChange, P.Value <= 0.05) %>% row.names(),
                             GCB = subset(GCB_FoldChange, P.Value <= 0.05) %>% row.names()),
                        fill_color = c("#5b9bd5","#ed7d31"), 
                        fill_alpha = 0.3,
                        stroke_color = "grey",
                        show_percentage = FALSE)
Sup1F1 <- Volcano::volcanoPlot(ABC_FoldChange, SigName = "P.Value", FCName = "logFC", fcThres = 2, pThres = 0.01, PRINT = FALSE)
Sup1F2 <- Volcano::volcanoPlot(GCB_FoldChange, SigName = "P.Value", FCName = "logFC", fcThres = 2, pThres = 0.01, PRINT = FALSE)
## G
Sup1G <- ggvenn::ggvenn(list(ABC = subset(ABC_FoldChange, P.Value <= 0.01 & abs(logFC) > 1) %>% row.names(),
                             GCB = subset(GCB_FoldChange, P.Value <= 0.01 & abs(logFC) > 1) %>% row.names()),
                        fill_color = c("#5b9bd5","#ed7d31"), 
                        fill_alpha = 0.3,
                        stroke_color = "grey",
                        show_percentage = FALSE)
```

# Load in the testing cohorts

```{r}
CutOff = 0
Predictor <- names(betas)
## Load Cohorts
REMoDLB <- readRDS("RDS/ExternalDataset/REMoDLB.Processed.ABC.rds")
HMRN <- readRDS("RDS/ExternalDataset/HMRN.Processed.rds")
LLMPP <- readRDS("RDS/ExternalDataset/LLMPP.Processed.ABC.RCHOP.rds")
```

## Generating Z score predictors
## Testing

```{r}
REMoDLB_Fil <- GenerateLinearPredictor(betas, REMoDLB)
REMoDLB_Fil$BestGroup <- ifelse(REMoDLB_Fil$LinearPredictor <= CutOff, "Low", "High")
REMoDLB_Fil$OS_TIME_Trunc <- ifelse(REMoDLB_Fil$OS_TIME > 3, 3, REMoDLB_Fil$OS_TIME)
REMoDLB_Fil$OS_IND_Trunc <- ifelse(REMoDLB_Fil$OS_TIME > 3, 0, REMoDLB_Fil$OS_IND)
# REMoDLB_Fil <- FindBestCuttoff(REMoDLB_Fil)
REMoDLB_Annot <- data.frame(x=c(4.5,4.5),
                            y=c(0.775,0.925),
                            x_trunc=c(2.75, 2.75),
                            y_trunc=c(0.7,0.9),
                            label=paste("n =", table(REMoDLB_Fil$BestGroup)))
REMoDLB_Full <- ggsurvplot(survfit(Surv(OS_TIME, OS_IND) ~ BestGroup,
                                   data=pData(REMoDLB_Fil)),
                           pval = TRUE,
                           pval.method = TRUE,
                           pval.size=4,
                           surv.median.line	= "hv",
                           ylab="Overall Survival",
                           xlab = "Follow up time (years)")$plot +
        geom_text(data=REMoDLB_Annot, aes(x=x, y=y, label=label), size=4) +
        theme(axis.text.x = element_text(size=10),
              axis.text.y = element_text(size=10),
              axis.title.x = element_text(size=12),
              axis.title.y = element_text(size=12)) +
        ## Box
        ggplot(pData(REMoDLB_Fil), 
               aes(y = LinearPredictor, x = BestGroup, fill = BestGroup)) +
        geom_boxplot()  +
        geom_hline(yintercept = 0, colour = "red", linetype = "dashed") +
        plot_layout(ncol = 2, widths = c(2,1)) &
        theme_classic() &
        theme(legend.position = "none")

REMoDLB_3year <- ggsurvplot(survfit(Surv(OS_TIME_Trunc, OS_IND_Trunc) ~ BestGroup,
                                    data=pData(REMoDLB_Fil)),
                            pval = TRUE,
                            pval.method = TRUE,
                            pval.size=4,
                            surv.median.line	= "hv",
                            ylab="Overall Survival",
                            xlab = "Follow up time (years)")$plot +
        geom_text(data=REMoDLB_Annot, aes(x=x_trunc, y=y_trunc, label=label), size=4) +
        theme(axis.text.x = element_text(size=10),
              axis.text.y = element_text(size=10),
              axis.title.x = element_text(size=12),
              axis.title.y = element_text(size=12)) +
        ## Box
        ggplot(pData(REMoDLB_Fil), 
               aes(y = LinearPredictor, x = BestGroup, fill = BestGroup)) +
        geom_boxplot()  +
        geom_hline(yintercept = 0, colour = "red", linetype = "dashed") +
        plot_layout(ncol = 2, widths = c(2,1)) &
        theme_classic() &
        theme(legend.position = "none")
```


```{r}
LLMPP_Fil <- GenerateLinearPredictor(betas, LLMPP)
LLMPP_Fil$BestGroup <- ifelse(LLMPP_Fil$LinearPredictor <= CutOff, "Low", "High")
LLMPP_Fil$OS_TIME_Trunc <- ifelse(LLMPP_Fil$OS_TIME > 3, 3, LLMPP_Fil$OS_TIME)
LLMPP_Fil$OS_IND_Trunc <- ifelse(LLMPP_Fil$OS_TIME > 3, 0, LLMPP_Fil$OS_IND)
# LLMPP_Fil <- FindBestCuttoff(LLMPP_Fil)
LLMPP_Annot <- data.frame(x=c(5,5),
                          y=c(0.5,0.75),
                          x_trunc=c(2.75, 2.75),
                          y_trunc=c(0.5,0.8),
                          label=paste("n =", table(LLMPP_Fil$BestGroup)))
LLMPP_Full <- ggsurvplot(survfit(Surv(OS_TIME, OS_IND) ~ BestGroup,
                                 data=pData(LLMPP_Fil)),
                         pval = TRUE,
                         pval.method = TRUE,
                         pval.size=4,
                         surv.median.line	= "hv",
                         ylab="Overall Survival",
                         xlab = "Follow up time (years)")$plot +
        geom_text(data=LLMPP_Annot, aes(x=x, y=y, label=label), size=4) +
        theme(axis.text.x = element_text(size=10),
              axis.text.y = element_text(size=10),
              axis.title.x = element_text(size=12),
              axis.title.y = element_text(size=12)) +
        ## Box
        ggplot(pData(LLMPP_Fil), 
               aes(y = LinearPredictor, x = BestGroup, fill = BestGroup)) +
        geom_boxplot()  +
        geom_hline(yintercept = 0, colour = "red", linetype = "dashed") +
        plot_layout(ncol = 2, widths = c(2,1)) &
        theme_classic() &
        theme(legend.position = "none")

LLMPP_3year <- ggsurvplot(survfit(Surv(OS_TIME_Trunc, OS_IND_Trunc) ~ BestGroup,
                                  data=pData(LLMPP_Fil)),
                          pval = TRUE,
                          pval.method = TRUE,
                          pval.size=4,
                          surv.median.line	= "hv",
                          ylab="Overall Survival",
                          xlab = "Follow up time (years)")$plot +
        geom_text(data=LLMPP_Annot, aes(x=x_trunc, y=y_trunc, label=label), size=4) +
        theme(axis.text.x = element_text(size=10),
              axis.text.y = element_text(size=10),
              axis.title.x = element_text(size=12),
              axis.title.y = element_text(size=12)) +
        ## Box
        ggplot(pData(LLMPP_Fil), 
               aes(y = LinearPredictor, x = BestGroup, fill = BestGroup)) +
        geom_boxplot()  +
        geom_hline(yintercept = 0, colour = "red", linetype = "dashed") +
        plot_layout(ncol = 2, widths = c(2,1)) &
        theme_classic() &
        theme(legend.position = "none")
```

```{r}
HMRN_Fil <- GenerateLinearPredictor(betas, HMRN)
HMRN_Fil$BestGroup <- ifelse(HMRN_Fil$LinearPredictor <= CutOff, "Low", "High")
HMRN_Fil$OS_TIME_Trunc <- ifelse(HMRN_Fil$OS_TIME > 3, 3, HMRN_Fil$OS_TIME)
HMRN_Fil$OS_IND_Trunc <- ifelse(HMRN_Fil$OS_TIME > 3, 0, HMRN_Fil$OS_IND)
# HMRN_Fil <- FindBestCuttoff(HMRN_Fil)
HMRN_Annot <- data.frame(x=c(12,12),
                         y=c(0.2,0.5),
                         x_trunc=c(2.75, 2.75),
                         y_trunc=c(0.55,0.85),
                         label=paste("n =", table(HMRN_Fil$BestGroup)))
HMRN_Full <- ggsurvplot(survfit(Surv(OS_TIME, OS_IND) ~ BestGroup,
                                data=pData(HMRN_Fil)),
                        pval = TRUE,
                        pval.method = TRUE,
                        pval.size=4,
                        log.rank.weights = "n",
                        surv.median.line	= "hv",
                        ylab="Overall Survival",
                        xlab = "Follow up time (years)")$plot +
        geom_text(data=HMRN_Annot, aes(x=x, y=y, label=label), size=4) +
        theme(axis.text.x = element_text(size=10),
              axis.text.y = element_text(size=10),
              axis.title.x = element_text(size=12),
              axis.title.y = element_text(size=12)) +
        ## Box
        ggplot(pData(HMRN_Fil), 
               aes(y = LinearPredictor, x = BestGroup, fill = BestGroup)) +
        geom_boxplot()  +
        geom_hline(yintercept = 0, colour = "red", linetype = "dashed") +
        plot_layout(ncol = 2, widths = c(2,1)) &
        theme_classic() &
        theme(legend.position = "none")

HMRN_3year <- ggsurvplot(survfit(Surv(OS_TIME_Trunc, OS_IND_Trunc) ~ BestGroup,
                                 data=pData(HMRN_Fil)),
                         pval = TRUE,
                         pval.method = TRUE,
                         pval.size=4,
                         surv.median.line	= "hv",
                         ylab="Overall Survival",
                         xlab = "Follow up time (years)")$plot +
        geom_text(data=HMRN_Annot, aes(x=x_trunc, y=y_trunc, label=label), size=4) +
        theme(axis.text.x = element_text(size=10),
              axis.text.y = element_text(size=10),
              axis.title.x = element_text(size=12),
              axis.title.y = element_text(size=12)) +
        ## Box
        ggplot(pData(HMRN_Fil), 
               aes(y = LinearPredictor, x = BestGroup, fill = BestGroup)) +
        geom_boxplot()  +
        geom_hline(yintercept = 0, colour = "red", linetype = "dashed") +
        plot_layout(ncol = 2, widths = c(2,1)) &
        theme_classic() &
        theme(legend.position = "none")
```

## Cause of death in HMRN cohort

```{r}
CauseOfDeath <- read.xlsx("HMRN/COD_2021_10_10.xlsx", sheetIndex = 1) %>%
    setNames(c("DASL_ID", "COD")) %>% 
    subset(! DASL_ID %>% duplicated())
HMRN_Fil_x <- HMRN_Fil

pData(HMRN_Fil) <- left_join(pData(HMRN_Fil), CauseOfDeath)


ggsurvplot(survfit(Surv(OS_TIME, OS_IND) ~ COD,
                   data=pData(HMRN_Fil)),
           pval = TRUE,
           pval.method = TRUE,
           pval.size=4,
           surv.median.line	= "hv",
           ylab="Overall Survival",
           xlab = "Follow up time (years)")

pData(HMRN_Fil) %>%
        subset(OS_IND == 1) %>%
        ggplot(aes(x = COD ,y = OS_TIME)) + 
        geom_boxplot(outlier.alpha = 0) + 
        stat_compare_means(label.x = 1.5, hjust = 0.5, label.y = 15) +
        labs(x = "Cause of death", y = "follow up time") +
        geom_jitter() +
        geom_hline(yintercept = 3, linetype = "dashed", colour = "red")
ggsave("Figures/Paper_Figures/HMRN.cause.of.death.png", width = 4, height = 3)
ggsave("Figures/Paper_Figures/HMRN.cause.of.death.svg", width = 4, height = 3)
table(HMRN_Fil$OS_TIME > 1, HMRN_Fil$COD)
table(HMRN_Fil$OS_TIME > 2, HMRN_Fil$COD)
table(HMRN_Fil$OS_TIME > 3, HMRN_Fil$COD)
table(HMRN_Fil$OS_TIME > 4, HMRN_Fil$COD)
table(HMRN_Fil$OS_TIME > 5, HMRN_Fil$COD)


42/(42+13)
6/(6+31)
```


# Testing cohort Figures (2 - 3 year, Supp 3 Full)

```{r}
## REMoDLB
## 3 year
REMoDLB_3year
ggsave("Figures/Paper_Figures/Figure2-REMoDLB.threeyear.svg", width = 7, height = 4)
## Full
REMoDLB_Full
ggsave("Figures/Paper_Figures/SupplementalFigure2-REMoDLB.svg", width = 7, height = 4)
## LLMPP
## 3 year
LLMPP_3year
ggsave("Figures/Paper_Figures/Figure2-LLMPP.threeyear.svg", width = 7, height = 4)
## Full
LLMPP_Full
ggsave("Figures/Paper_Figures/SupplementalFigure2-LLMPP.svg", width = 7, height = 4)
## HMRN
## 3 year
HMRN_3year
ggsave("Figures/Paper_Figures/Figure2-HMRN.threeyear.svg", width = 7, height = 4)
## Full
HMRN_Full
ggsave("Figures/Paper_Figures/SupplementalFigure2-HMRN.svg", width = 7, height = 4)
```

## Survival Stats for the paper

```{r eval = FALSE}
## Data for the paper:
## 3 year survival statistics
coxph(Surv(OS_TIME_Trunc, OS_IND_Trunc) ~ (BestGroup == "High"),
      data=pData(REMoDLB_Fil)) %>%
        summary()
coxph(Surv(OS_TIME_Trunc, OS_IND_Trunc) ~ (BestGroup == "High"),
      data=pData(LLMPP_Fil)) %>%
        summary()
coxph(Surv(OS_TIME_Trunc, OS_IND_Trunc) ~ (BestGroup == "High"),
      data=pData(HMRN_Fil)) %>%
        summary()
## Full time stats
coxph(Surv(OS_TIME, OS_IND) ~ (BestGroup == "High"),
      data=pData(REMoDLB_Fil)) %>%
        summary()
coxph(Surv(OS_TIME, OS_IND) ~ (BestGroup == "High"),
      data=pData(LLMPP_Fil)) %>%
        summary()
coxph(Surv(OS_TIME, OS_IND) ~ (BestGroup == "High"),
      data=pData(HMRN_Fil)) %>%
        summary()
```

## Multivariate analysis

```{r}

coxph(Surv(OS_TIME, OS_IND) ~ (BestGroup == "High") +
              as.numeric(AGE) + 
              (GENDER == "Male") + 
              (IPI_SCORE > 2) + 
              (grepl("III|IV", STAGE)),
                    data=pData(REMoDLB_Fil)) %>% summary()


pData(LLMPP)
rbind(cbind(Dataset = "REMoDLB",
      coxph(Surv(OS_TIME, OS_IND) ~ (BestGroup == "High") + 
              as.numeric(AGE) + 
              (GENDER == "Male") + 
              (IPI_SCORE > 2) + 
              (grepl("III|IV", STAGE)),
                    data=pData(REMoDLB_Fil)) %>%
                      summary() %>% 
                      .$coefficients %>% 
                      as.data.frame()),
        cbind(Dataset = "LLMPP", 
              coxph(Surv(OS_TIME, OS_IND) ~ 
                            as.numeric(`Age:ch1`) + 
                            `Gender:ch1` + 
                            (BestGroup == "High"),
                    data=pData(LLMPP_Fil)) %>%
                      summary() %>% 
                      .$coefficients %>% 
                      as.data.frame())) %>%
        xlsx::write.xlsx("SpreadSheets/MultiVariate.xlsx", "Sheet1")
```

## Genomic Testing

```{r}
ColoursGenomic <- c(
        ## Same Groups
        "MCD" = "firebrick",
        "MYD88" = "firebrick",
        ## Mix of MCD and ST2
        "MCD/ST2" = "red",
        ## Same Groups
        "BCL2" = "goldenrod",
        "EZB" = "goldenrod",
        ## Same Groups
        "BN2" = "chartreuse3",
        "NOTCH2" = "chartreuse3",
        ## No match
        "ST2" = "orchid3",
        ## No match
        "N1" = "skyblue",
        ## No match
        "SOCS1/SGK1" = "gold",
        ## No match
        "TET2/SGK1" = "slateblue3",
        ## Both have others make them gray
        "Other" = "gray83")
pData(HMRN_Fil) %>% 
        subset(! is.na(Lacy_class)) %>%
        select(BestGroup,Lacy_class) %>%
        ggplot(aes(x = gsub(" Risk", "", BestGroup), fill = Lacy_class)) +
        geom_bar(position = "fill") +
        scale_fill_manual(values=ColoursGenomic[levels(as.factor(HMRN$Lacy_class))], 
                          guide = guide_legend(title.position = "top",
                                               title.hjust = 0.5,
                                               title.theme = element_text(size=8),
                                               label.theme = element_text(size=6),
                                               keywidth = 0.5,
                                               keyheight = 0.5),
                          drop = TRUE) +
        scale_y_continuous(labels = scales::percent) +
        labs(x = "Risk Group", y = "") +
        theme_classic() +
        theme(legend.position = "bottom",
              legend.title = element_text(size=8), 
              legend.text=element_text(size=8),legend.key.size = unit(4, "mm"))
ggsave("Figures/Paper_Figures/SupplementalFigure3_Genomics.svg", height = 4, width = 2)

table(is.na(HMRN_Fil$Lacy_class))
table(HMRN_Fil$BestGroup, HMRN_Fil$Lacy_class) %>% fisher.test()
table(HMRN_Fil$BestGroup, HMRN_Fil$Lacy_class)
table(HMRN_Fil$BestGroup)

  
pData(REMoDLB_Fil) %>% 
    ggplot(aes(x = BestGroup, fill = as.character(IPI_SCORE))) +
    geom_bar(position = "fill") + 
    scale_y_continuous(labels = scales::percent) +
    labs(x = "Risk Group", y = "", fill = "IPI")+
    theme_classic() +
    theme(legend.position = "bottom",
          legend.title = element_text(size=8), 
          legend.text=element_text(size=8),legend.key.size = unit(4, "mm"))
ggsave("Figures/Paper_Figures/SupplementalFigure3_IPI.svg", height = 4, width = 2)
table(REMoDLB_Fil$IPI, REMoDLB_Fil$BestGroup) %>% fisher.test()
```

## Random

### Load iterations and combine

```{r eval  = FALSE}
random <- list()
random$Stats <- lapply(list.files("RDS/Random/", full.names = TRUE), function(file) {
        RandIter <- readRDS(file)
        lapply(RandIter, function(itr) {itr$stats}) %>% do.call("rbind",.) %>% as.data.frame()
}) %>% do.call("rbind",.) %>% remove_rownames()
random$Betas <- lapply(list.files("RDS/Random/", full.names = TRUE), function(file) {
        RandIter <- readRDS(file)
        lapply(RandIter, function(itr) {unname(itr$betas)}) %>% do.call("rbind",.) %>% as.data.frame()
}) %>% do.call("rbind",.) %>% remove_rownames()
random$Genes <- lapply(list.files("RDS/Random/", full.names = TRUE), function(file) {
        RandIter <- readRDS(file)
        lapply(RandIter, function(itr) {names(itr$betas)}) %>% do.call("rbind",.) %>% as.data.frame()
}) %>% do.call("rbind",.) %>% remove_rownames()
saveRDS(random, "RDS/Random/Random_all.zscore.rds")
```

### Test Numbers

```{r}
random <- readRDS("RDS/Random/Random_all.zscore.rds")
x <- coxph(Surv(OS_TIME, OS_IND) ~ BestGroup == "High",data=pData(REMoDLB_Fil))
REMoDLB_HR <- exp(x$coefficients)
y <- summary(x)
REMoDLB_P <- y$sctest[3]
table(random$Stats$REMoDLB_HR >= REMoDLB_HR)
100 - 13491/300000 * 100
table(random$Stats$REMoDLB_P <= REMoDLB_P)
100 - 20164/300000 * 100
## LLMPP
x <- coxph(Surv(OS_TIME, OS_IND) ~ BestGroup == "High",data=pData(LLMPP_Fil))
LLMPP_HR <- exp(x$coefficients)
y <- summary(x)
LLMPP_P <- y$sctest[3]
table(random$Stats$LENZ_HR >= LLMPP_HR)
100- 5029 /300000 * 100
table(random$Stats$LENZ_P <= LLMPP_P)
100 - 6705/300000 * 100
table(random$Stats$LENZ_HR >= LLMPP_HR & random$Stats$REMoDLB_HR >= REMoDLB_HR)
100 - 227/299773 * 100
OurCox <- rbind(cbind("LLMPP", LLMPP_HR %>% unname(), 294971),
                cbind("REMoDL-B", REMoDLB_HR %>% unname(), 286509)) %>%
        as.data.frame() %>%
        setNames(c("variable", "HR","greaterthan"))

```

### Visualise

```{r}
random$Stats %>% 
        reshape2::melt() %>% 
        subset(grepl("HR", variable)) %>% 
        mutate(variable = variable %>% gsub("LENZ_HR", "LLMPP", .) %>% gsub("REMoDLB_HR","REMoDL-B",.) ) %>%
        arrange(value) %>%
        mutate(iter = uniqtag::cumcount(variable %>% as.character())) %>%
        ggplot(aes(x = iter, y = value)) + 
        geom_point() + 
        geom_hline(data = OurCox, aes(yintercept = HR %>% as.numeric()), linetype = "dashed") +
        geom_vline(data = OurCox, aes(xintercept = greaterthan %>% as.numeric()), linetype = "dashed") +
        facet_wrap(~variable,nrow =  2) +
        labs(y = "HR", x = "Iteration") +
        theme_classic() +
        theme(strip.background = element_blank())
ggsave("Figures/Paper_Figures/SupplementalFigure3_RandomGeneSets.png", width = 8, height = 4)
ggsave("Figures/Paper_Figures/SupplementalFigure3_RandomGeneSets.svg", width = 8, height = 4)
```




